BackendServer/python/analysis.py

```diff
@@ -0,0 +1,163 @@
+    output_df['Assumed_Link_Points'] = 0
+
+    # Calculate mobility/station points
+    output_df['Mobility_Points'] = output_df['Mobility'].map({0: 0, 1: 3})
+    output_df['Auto_Station_Points'] = output_df['Auto_Station'].map({AutoStation.noPoints: 0, AutoStation.failed: 0, AutoStation.docked: 8, AutoStation.engaged: 12})
+    output_df['Tele_Station_Points'] = output_df['Tele_Station'].map({TeleStation.noPoints: 0, TeleStation.parked: 0, TeleStation.failed: 0, TeleStation.docked: 8, TeleStation.engaged: 12})
+
+    # Total points
+    output_df['Total_Auto_Points'] = output_df['Auto_Grid_Points'] + output_df['Auto_Station_Points'] + output_df['Mobility_Points']
+    output_df['Total_Tele_Points'] = output_df['Tele_Grid_Points'] + output_df['Tele_Station_Points'] + output_df['Assumed_Link_Points']
+    output_df['Total_Points'] = output_df['Total_Auto_Points'] + output_df['Total_Tele_Points']
+    
+    return output_df
+
+def calculate_match_analysis(Team_Number, db_connection, *, appendTo = {}):
+    # Read data from MySQL database
+    matches_df = pd.read_sql('SELECT * FROM matchData WHERE Team_Number = %s AND No_Show_Robot = FALSE', db_connection, params=(Team_Number,))
+
+    # Calculate additional scores
+    calculate_match_scores(matches_df, mutate=True)
+    
+    # Create output variable
+    analysis_output = appendTo
+    
+    # Insert team number
+    analysis_output['Team_Number'] = Team_Number
+
+    # Calculate scoring aggregates
+    for type_phase, prefixes in (
+        ('Auto',        ('Auto_Cone', 'Auto_Cube')),
+        ('Tele_Pieces', ('Tele_Cone', 'Tele_Cube')),
+        ('Tele_Cone',   ('Tele_Cone',)),
+        ('Tele_Cube',   ('Tele_Cube',))
+    ):
+        for level, suffixes in (
+            ('Total', ('Low', 'Mid', 'High')),
+            ('Low',   ('Low',)),
+            ('Mid',   ('Mid',)),
+            ('High',  ('High',))
+        ):
+            # Get names of all the columns we want
+            included_columns = [f"{prefix}_{suffix}" for prefix in prefixes for suffix in suffixes]
+
+            # Sum up all of the included columns, then find the mean or max of it
+            analysis_output[f'{type_phase}_{level}_Average'] = matches_df[included_columns].sum(axis=1).mean()
+            analysis_output[f'{type_phase}_{level}_Max'] =     matches_df[included_columns].sum(axis=1).max()
+
+    # Calculate station frequencies
+    tele_station_attempts = (matches_df['Tele_Station'].isin(TELE_ATTEMPTED_DOCK_STATES)).sum()
+    analysis_output['End_Balance_Frequency'] = div((matches_df['Tele_Station'] == TeleStation.engaged).sum(), tele_station_attempts)
+    analysis_output['End_Dock_Frequency'] = div((matches_df['Tele_Station'].isin(TELE_SUCCESSFUL_DOCK_STATES)).sum(), tele_station_attempts)
```

Comment:
Is it true that `End_Dock_Frequency` will represent the percentage of matches a team was either able to dock or engage?

Reply:
Yes, out of all the matches they attempted